File: archivist/src/main/java/com/zorroa/archivist/domain/Permission.java
package com.zorroa.archivist.domain;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.google.common.base.MoreObjects;
import org.springframework.security.core.GrantedAuthority;

import java.io.Serializable;
import java.util.Objects;
import java.util.UUID;

/**
 * Created by chambers on 10/27/15.
 */
public class Permission implements Loggable<UUID>, Serializable, GrantedAuthority {

    public static final String JOIN = "::";

    private UUID id;
    private String name;
    private String type;
    private String description;
    private boolean immutable;

    public String getFullName() {
        return type.concat(JOIN).concat(name);
    }

    public UUID getId() {
        return id;
    }

    public Permission setId(UUID id) {
        this.id = id;
        return this;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public boolean isImmutable() {
        return immutable;
    }

    public Permission setImmutable(boolean immutable) {
        this.immutable = immutable;
        return this;
    }

    @JsonIgnore
    public String getAuthority() {
        return type.concat(JOIN).concat(name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, 829);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Permission that = (Permission) o;
        return Objects.equals(getId(), that.getId());
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(getClass())
                .add("id", id)
                .add("name", name)
                .add("type", type)
                .add("description", description)
                .toString();
    }

    @Override
    public UUID getTargetId() {
        return id;
    }
}

File: archivist/src/main/kotlin/com/zorroa/archivist/domain/Permission.kt
package com.zorroa.archivist.domain

import com.fasterxml.jackson.annotation.JsonIgnore
import com.zorroa.archivist.security.getOrgId
import com.zorroa.common.repository.KDaoFilter
import io.swagger.annotations.ApiModel
import io.swagger.annotations.ApiModelProperty
import org.springframework.security.core.GrantedAuthority
import java.util.UUID

@ApiModel("Set Permissions", description = "Directions for setting a Permission.")
class SetPermissions constructor() {

    @ApiModelProperty("ACL to set.")
    var acl: Acl? = null

    @ApiModelProperty("If true the Permissions will be replaced instead of appended.")
    var replace = false

    constructor(acl: Acl?, replace: Boolean) : this() {
        this.acl = acl
        this.replace = replace
    }
}

class PermissionUpdateSpec(
    val id: UUID,
    val type: String,
    val name: String,
    val description: String
)

@ApiModel("Permission Spec", description = "Attributes required to create a Permission.")
class PermissionSpec(

    @ApiModelProperty("Type of the Permission.")
    var type: String,

    @ApiModelProperty("Name of the Permission.")
    var name: String,

    @ApiModelProperty("Description of the Permission.")
    var description: String = "",

    @ApiModelProperty("Source of the Permission.")
    var source: String = "local"

) {

    constructor(authority: String) : this(authority.split(Permission.JOIN)[0],
            authority.split(Permission.JOIN)[1])
}

/**
 * A Permission describes a group or role.  Permission implements from GrantedAuthority which
 * allows it to be used by Spring directly.
 */
@ApiModel("Permission", description = "Describes a user Permission.")
class Permission constructor (

    @ApiModelProperty("UUID of the Permission.")
    val id: UUID,

    @ApiModelProperty("Name of the Permission.")
    val name: String,

    @ApiModelProperty("Type of the Permission.")
    val type: String,

    @ApiModelProperty("Description of the Permission.")
    val description: String,

    @JsonIgnore
    val immutable: Boolean = false

) : GrantedAuthority {

    val fullName = "$type${Permission.JOIN}$name"

    override fun getAuthority(): String {
        return fullName
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Permission

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    override fun toString() = authority

    companion object {
        /**
         * A string for joining together the type + name
         */
        const val JOIN = "::"
    }
}

class PermissionFilter constructor(
    val types: List<String>? = null,
    val names: List<String>? = null,
    val authorities: List<String>? = null,
    val assignableToUser: Boolean? = null,
    val assignableToObj: Boolean? = null
) : KDaoFilter() {

    @JsonIgnore
    override val sortMap: Map<String, String> =
            mapOf("id" to "permission.pk_permission",
                    "name" to "permission.str_name",
                    "type" to "permission.str_type",
                    "authority" to "permission.str_authority")

    override fun build() {
        if (sort == null) {
            sort = listOf("authority:a")
        }

        addToWhere("permission.pk_organization=?")
        addToValues(getOrgId())

        types?.let {
            addToWhere(com.zorroa.common.util.JdbcUtils.inClause("permission.str_type", it.size))
            addToValues(it)
        }

        names?.let {
            addToWhere(com.zorroa.common.util.JdbcUtils.inClause("permission.str_name", it.size))
            addToValues(it)
        }

        authorities?.let {
            addToWhere(com.zorroa.common.util.JdbcUtils.inClause("permission.str_authority", it.size))
            addToValues(it)
        }

        assignableToUser?.let {
            addToWhere("bool_assignable_to_user=?")
            addToValues(assignableToUser)
        }

        assignableToObj?.let {
            addToWhere("bool_assignable_to_obj=?")
            addToValues(assignableToObj)
        }
    }
}
